### DISCUSSION TITLE
Is there a way to save a pdf to a specific PDF version?

### DISCUSSION THREAD

**Discussion by cmyers009** [pymupdf/PyMuPDF#3348] [Looking for help] [answered]
1.7  is the PDF version I get when I save the pdf with default parameters.

If I wanted to save the PDF as version 1.6 or 1.5 for example, is this possible for me to do?
I looked through the documentation and could not find any reference to this.

**Comment by cmyers009** [comment, marked as answer]
Found my answer - Document metadata.
https://pymupdf.readthedocs.io/en/latest/document.html#Document.metadata

**Comment by JorjMcKie** [comment]
The PDF version in `Document.metadata` is info-only.
You cannot modify this respectively convert a PDF to a different version.

**Comment by cbm755** [comment]
I'm having trouble where users:
  - produce a PDF-1.5 file that works on their printer
  - then we do some minimal manipulation with PyMuPDF (adding images, small amounts of text).  The result is PDF-1.7
  - users complain that they cannot print.

I'm sure its poor printer drivers etc but I'd like to be able to rule out PyMuPDF.  Is it on the road map to force saving to a lower version during save, say `Document.save(..., format="pdf-1.5")`?

**Reply by JorjMcKie** [reply]
No, we always save **_new_** files in the most current format.
But I am pretty certain, that the PDF format has no influence on the error.
Did you try / consider saving incrementally? That would not change the PDF version. If insertions are the only things happening (and no updates like redactions, deletions, ...), there would not any drawback.
I also hope you guys don't use `page.clean_contents()` anymore btw.

On another note, I just took a PDF-1.5 file, inserted text, then saved via `.ez_save()` ... this does not change the PDF version.
So I am wondering what your users **_really_** are doing.

**Reply by cbm755** [reply]
> I also hope you guys don't use page.clean_contents() anymore btw.

Oh!  I found we do!
```diff
     for p in range(1, spec["numberOfPages"] + 1):
-        # Workaround Issue #1347: unnecessary for pymupdf>=1.18.7
-        doc[p - 1].clean_contents()
```
I shall remove that, and see if users see the same problem.

- - - -
 
I also realize the main disruptive thing we do is separate PDF into into pages and reassemble.  In some cases we must do that (interleaves pages from 2 or more files).  But in the most-common case of "one version" we still use `doc.insert_pdf(...)` to takes the pages from the source into a new document.  Anyway, I'll get rid of the `clean_contents()` and see if that helps before I worry more about the `insert_pdf`.

**Reply by cbm755** [reply]
We save like this:
```
    doc.save(save_name, garbage=4, deflate=True, clean=True)
```

**Reply by JorjMcKie** [reply]
The most size-efficient way to save while keeping a good balance with performance is using `doc.ez_save(...)`.
I chose that name for phonetic reasons - sounds like "easy save".

It is short for `doc.save(filename, garbage=3, deflate=True, use_objstms=True)`. The `use_objstms` option often has a very significant effect on the resulting file size, because the object definitions (which normaly are uncompressed ASCII text) are collectively ZIP-compressed in chunks of streams.

Using `doc.ez_save(filename, garbage=4)` is possible and has a **_small_** chance to squeeze out a little more grease. In contrast to this, a significantly longer duration is **_guaranteed_**. Except in special situations, its use is generally discouraged.

Using `clean=True` has the same effect as executing `page.clean_contents()` for all pages and has **_no beneficial effect_** on the resulting file size (modulo epsilon 😉).

**Reply by cbm755** [reply]
Currently, I'm less concerned with file size and mostly concerned with correctness of the resulting PDF...  Is `ez_save()` with no additional options recommended in that case too?

**Reply by cbm755** [reply]
correctness is probably the wrong word, sorry.  "broad compatibility with printers shipped with ancient questionable software" is closer to what I mean.

**Reply by JorjMcKie** [reply]
Hm, hard to tell the relationship between the computer preparing the data stream it sends to the printer and the printer's hardware or its "software" (e-prom).
On the safe side with good performance definitely is `.save(filename, garbage=3, deflate=True)`.
